cirq/experiments/fidelity_estimation_test.py

```diff
@@ -437,3 +438,100 @@ def test_characterize_phased_fsim_parameters_with_xeb():
+    circuit_i = task['circuit_i']
+    cycle_depth = task['cycle_depth']
+    circuit = task['circuit']
+    param_resolver = task['param_resolver']
+
+    circuit_depth = cycle_depth * 2 + 1
+    assert circuit_depth <= len(circuit)
+    tcircuit = circuit[:circuit_depth]
+    tcircuit = cirq.resolve_parameters_once(tcircuit, param_resolver=param_resolver)
+
+    pure_sim = cirq.Simulator()
+    psi = cast(cirq.StateVectorTrialResult, pure_sim.simulate(tcircuit))
+    psi = psi.final_state_vector
+    pure_probs = np.abs(psi) ** 2
+
+    return {
+        'circuit_i': circuit_i,
+        'cycle_depth': cycle_depth,
+        'pure_probs': pure_probs,
+    }
+
+
+def _ref_simulate_2q_xeb_circuits(
+    circuits: Sequence['cirq.Circuit'],
+    cycle_depths: Sequence[int],
+    param_resolver: 'cirq.ParamResolverOrSimilarType' = None,
+    pool: Optional['multiprocessing.pool.Pool'] = None,
+):
+    """Reference implementation for `simulate_2q_xeb_circuits` that
+    does each circuit independently instead of using intermediate states.
+
+    You can also try editing the helper function to use QSimSimulator() for
+    benchmarking. This simulator does not support intermediate states, so
+    you can't use it with the new functionality.
+    https://github.com/quantumlib/qsim/issues/101
+    """
+    tasks = []
+    for cycle_depth in cycle_depths:
+        for circuit_i, circuit in enumerate(circuits):
+            tasks += [
+                {
+                    'circuit_i': circuit_i,
+                    'cycle_depth': cycle_depth,
+                    'circuit': circuit,
+                    'param_resolver': param_resolver,
+                }
+            ]
+
+    if pool is not None:
+        records = pool.map(_ref_simulate_2q_xeb_circuit, tasks, chunksize=4)
+    else:
+        records = [_ref_simulate_2q_xeb_circuit(record) for record in tasks]
+
+    return pd.DataFrame(records).set_index(['circuit_i', 'cycle_depth']).sort_index()
+
+
+def test_incremental_simulate():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuits = [
+        rqcg.random_rotations_between_two_qubit_circuit(
+            q0,
+            q1,
+            depth=100,
+            two_qubit_op_factory=lambda a, b, _: SQRT_ISWAP(a, b),
+        )
+        for _ in range(20)
+    ]
+    cycle_depths = np.arange(3, 100, 9)
+    pool = multiprocessing.Pool()
+
+    with multiprocessing.Pool() as pool:
+        start = time.perf_counter()
+        df_ref = _ref_simulate_2q_xeb_circuits(
+            circuits=circuits,
+            cycle_depths=cycle_depths,
+            pool=pool,
+        )
+        end1 = time.perf_counter()
+
+        df = simulate_2q_xeb_circuits(circuits=circuits, cycle_depths=cycle_depths, pool=pool)
+        end2 = time.perf_counter()
+
+    # uncomment for benchmarks:
+    # print("\nnew:", end2 - end1, "old:", end1 - start)
+
+    pd.testing.assert_frame_equal(df_ref, df)
+
+    # Use below for approximate equality, if e.g. you're using qsim:
```

Comment:
`pd.testing.assert_frame_equal` doesn't work if you're using e.g. `QSimSimulator` because of floating point comparison issues. The commented out code does approximate comparison of the dataframe arrays